### ISSUE TITLE
ep=1 and ep=8 use the same amount of GPU memory

### ISSUE THREAD

**Issue by imhuim982** [ByteDance-Seed/VeOmni#69]
Thank you for your great work! This is an outstanding open-source training framework. We are trying to train a MoE model, and noticed that when setting ep=1 or ep=8, the GPU memory usage remains the same. If parameter sharding is not supported, the usefulness of the ep setting seems limited. Is there any way to enable sharding for the moe parameters as well? We would greatly appreciate it if you could help support this.

**Comment by imhuim982** [comment]
I found in torch_parallelize.py:

from line 250:
```
if fsdp_no_shard_states is not None:
          # apply NO_SHARD the ignored_states, but wrap into DDP
          if parallel_state.ep_fsdp_mesh["ep_fsdp"].size() == 1:  <-- is always 1 even I set expert_parallel_size = 8
              moe_sharding_strategy = ShardingStrategy.NO_SHARD
              ep_fsdp_device_mesh = parallel_state.fsdp_mesh
          else:
              moe_sharding_strategy = ShardingStrategy.FULL_SHARD
              ep_fsdp_device_mesh = parallel_state.ep_fsdp_mesh["ep_fsdp"]
          logger.info_rank0(f"Apply {moe_sharding_strategy} states on '{fsdp_no_shard_states_fqn}'.")
          fsdp_kwargs.pop("ignored_states", None)
          fsdp_kwargs.pop("auto_wrap_policy", None)
          fsdp_kwargs["sharding_strategy"] = moe_sharding_strategy
          fsdp_kwargs["device_mesh"] = ep_fsdp_device_mesh
          ....
```

**Comment by Luosuu** [comment]
Hi, would you like to try EP+FSDP2 again as we recently made a major upgrade?

**Comment by imhuim982** [comment]
@Luosuu Many thanks for your reply! The release of 0.1.0 is quite exciting and I tried FSDP2 + EP right away. But I found the previous  dependencies like torch 2.5.1 are not compatible with these relese. So I'm still waiting for the latest dependencies.

**Comment by Luosuu** [comment]
I see. Internally we are using torch 2.7.1 for deployment, you can try upgrade torch to 2.7.1 and it should be fine.

We will look into a more reliable method like uv to manege dependencies soon

**Comment by imhuim982** [comment]
Thank you very much! I'll give it a try.

**Comment by imhuim982** [comment]
> I see. Internally we are using torch 2.7.1 for deployment, you can try upgrade torch to 2.7.1 and it should be fine.
> 
> We will look into a more reliable method like uv to manege dependencies soon

@Luosuu  Sorry for bothering. I successfully ran MoE with FSDP2 + ep with torch 2.7.1. But I still found the memory footprint increases with the EP.  
Here are the speed and memory records for different settings.

fsdp2 ep=1 16K, bz 1, w/o gdc: 1.66s/it 57G
fsdp2 ep=2 16K, bz 1, w/o gdc: 2.02s/it 60G
fsdp2 ep=4 16K, bz 1, w/o gdc: 2.44s/it 70G

Are the results expected?

**Comment by Luosuu** [comment]
Hi @imhuim982 , could you please share more info about your job config? Thank you. 

I majorly test against Qwen3-MoE models. 

We set up CI and add more tests for other models soon after [PR#98](https://github.com/ByteDance-Seed/VeOmni/pull/98)

Thank you very much for your understanding

**Comment by imhuim982** [comment]
My script looks like:

```
bash train.sh tasks/train_torch.py configs/pretrain/deepseek.yaml \
    --model.config_path configs/model_configs/deepseek/DeepseekV3_16B.json \
    --model.tokenizer_path tokenizer_path \
    --data.train_path data_set_path \
    --data.datasets_type tokenized \
    --data.max_seq_len 8192 \
    --data.train_size 10000000000 \
    --data.data_type tokenized \
    --train.init_device meta \
    --train.data_parallel_mode fsdp2 \
    --train.use_wandb false \
    --train.rmpad false \
    --train.rmpad_with_pos_ids false \
    --train.enable_gradient_checkpointing true \
    --train.enable_mixed_precision \
    --train.micro_batch_size 2 \
    --train.global_batch_size 32 \
    --train.lr 2e-3 \
    --train.expert_parallel_size 1 \
    --train.ulysses_parallel_size 1 \
```

I ran it on a single machine with 8-GPUs. 
I'm not sure if I understand correctly—does EP reduce GPU memory usage only work when the world size is greater than 8? Otherwise, are all parameters still evenly distributed across 8 GPUs? And when there are more than 8 GPUs, does each node only store the parameters of the experts it is responsible for?

**Comment by Luosuu** [comment]
I think EP should save memory usage even with only 8 GPUs.

Could you please share your custom DeepseekV3_16B.json config so that I can quickly debug on my side? Thank you

**Comment by imhuim982** [comment]
In this case, I just used the official DeepSeek v3 config in VeOmni.

**Comment by Luosuu** [comment]
Hi @imhuim982, thank you for the follow up. I will look into it soon.

Just in case of any misunderstanding, in terms of the static memory consumption (if you measure the memory usage after each step), EP is expected be the same with FSDP2. But EP should have less peak memory usage than FSDP2.

**Comment by Luosuu** [comment]
@imhuim982 Hi, sharing a follow up on this issue

It seems to be a bug with PyTorch FSDP2 which triggers duplicated all-gather for experts when there is a shared expert computation after we compute the experts. We have a workaround here to be megred soon:

https://github.com/ByteDance-Seed/VeOmni/pull/173

in which we workaround this issue by computing shared expert first then experts.

**Comment by Luosuu** [comment]
Closing this issue since the fix has been merged. Feel free to reopen
